File: client/src/types/quiz.ts
import type { PlayableCodingTask } from '../../../shared/coding-catalog';
import type { PublicItemReview } from '../../../shared/curation';
export type CategoryType = 'react' | 'typescript' | 'git' | 'javascript' | 'nodejs' | 'nextjs' | 'html' | 'css' | 'dsa' | 'algorithms' | 'abbreviations' | 'general' | 'ai' | 'cool-stuff' | 'databases' | 'system-design' | 'testing' | 'devops' | 'security' | 'dev-world' | 'code-snippets' | 'continents' | 'capitals' | 'flags' | 'landforms' | 'climate' | 'population' | 'political' | 'economic' | 'cartography' | 'earth' | 'arithmetic' | 'fractions' | 'prealgebra' | 'algebra' | 'geometry' | 'trigonometry' | 'statistics' | 'precalculus' | 'calculus' | 'linear-algebra' | 'prehistory' | 'ancient' | 'classical' | 'medieval' | 'renaissance' | 'earlymodern' | 'industrial' | 'worldwars' | 'coldwar' | 'modern' | 'openings' | 'tactics' | 'strategy' | 'endgames' | 'combinations' | 'discrete-math' | 'number-theory' | 'multivariable-calculus' | 'differential-equations' | 'real-analysis' | 'geomorphology' | 'oceanography' | 'biogeography' | 'geopolitics' | 'gis' | 'historiography' | 'history-of-science' | 'economic-history' | 'intellectual-history' | 'military-history' | 'cell-biology' | 'skeletal-system' | 'muscular-system' | 'nervous-system' | 'endocrine-system' | 'cardiovascular-system' | 'respiratory-system' | 'digestive-system' | 'immune-system' | 'reproductive-system' | 'opening-theory' | 'middlegame' | 'pawn-structures' | 'endgame-technique' | 'chess-history' | 'positions' | 'starting-hands' | 'pot-odds' | 'betting-strategy' | 'postflop' | 'tournament-play' | 'psychology' | 'gto-advanced';

export interface Question {
  id: string;
  tags: string[];
  introduction: string;
  question: string;
  options: string[];
  category: CategoryType;
  difficulty: 1 | 2 | 3 | 4 | 5;
  /** The version of this exact wording, and whatever review is recorded for
   * it. Absent when the server could not resolve it, which is a reason to say
   * nothing about review rather than to assume the friendly answer. */
  review?: PublicItemReview;
}

export type DifficultyMode = 'basics' | 'easy' | 'zero-to-hero' | 'advanced' | 'mixed';

export interface QuizResult {
  totalQuestions: number;
  correctAnswers: number;
  percentage: number;
  questXp: number;
  /** Questions retired between the session being issued and the answers
   * arriving. Graded as void — neither for nor against — and left out of the
   * total, so the result screen says why the count is short. */
  voided?: string[];
  resultReceipt?: string;
  results: {
    questionId: string;
    selectedIndex: number;
    correctAnswer: number;
    isCorrect: boolean;
    explanation: string;
    scoreProof?: string;
    answerProof?: string;
  }[];
}

export type QuizState = 'loading' | 'ready' | 'in-progress' | 'submitted' | 'error';

/* ──── Roadmap ("Learn") mode ──────────────────────────────────────────── */

export type RoadmapTopic =
  | 'javascript' | 'typescript' | 'react' | 'nextjs' | 'nodejs'
  | 'html' | 'css' | 'git' | 'dsa' | 'algorithms'
  | 'abbreviations' | 'general' | 'ai' | 'cool-stuff'
  | 'databases' | 'system-design' | 'testing' | 'devops' | 'security'
  // Geography
  | 'continents' | 'capitals' | 'flags' | 'landforms' | 'climate' | 'population' | 'political' | 'economic' | 'cartography' | 'earth'
  // Math
  | 'arithmetic' | 'fractions' | 'prealgebra' | 'algebra' | 'geometry' | 'trigonometry' | 'statistics' | 'precalculus' | 'calculus' | 'linear-algebra'
  // History
  | 'prehistory' | 'ancient' | 'classical' | 'medieval' | 'renaissance' | 'earlymodern' | 'industrial' | 'worldwars' | 'coldwar' | 'modern'
  // Chess
  | 'openings' | 'tactics' | 'strategy' | 'endgames' | 'combinations' | 'discrete-math' | 'number-theory' | 'multivariable-calculus' | 'differential-equations' | 'real-analysis' | 'geomorphology' | 'oceanography' | 'biogeography' | 'geopolitics' | 'gis' | 'historiography' | 'history-of-science' | 'economic-history' | 'intellectual-history' | 'military-history' | 'cell-biology' | 'skeletal-system' | 'muscular-system' | 'nervous-system' | 'endocrine-system' | 'cardiovascular-system' | 'respiratory-system' | 'digestive-system' | 'immune-system' | 'reproductive-system' | 'opening-theory' | 'middlegame' | 'pawn-structures' | 'endgame-technique' | 'chess-history' | 'positions' | 'starting-hands' | 'pot-odds' | 'betting-strategy' | 'postflop' | 'tournament-play' | 'psychology' | 'gto-advanced';

export interface RoadmapLevelMeta {
  level: number;
  title: string;
  difficulty: 1 | 2 | 3 | 4 | 5;
  questionCount: number;
  /** devShark code topics only: how many coding tasks close this level. */
  codingTasks?: number;
  /** Too few reviewed questions to open. Stays on the map with its number;
   * never a prerequisite, and the rules step over it. */
  unavailable?: true;
}

export interface RoadmapCheckpointMeta {
  checkpoint: number;
  title: string;
  /** The level this checkpoint sits after (5, 10, 15, 20, 25). */
  afterLevel: number;
  questionCount: number;
  passPct: number;
}

export interface RoadmapTopicStructure {
  levels: RoadmapLevelMeta[];
  checkpoints: RoadmapCheckpointMeta[];
  /** Part tests with no available level to examine. */
  unavailableParts?: number[];
}

export interface RoadmapStructure {
  topics: RoadmapTopic[];
  structure: Record<RoadmapTopic, RoadmapTopicStructure>;
}

// A playable roadmap question. Correct answers stay server-side until this
// specific question has been submitted.
export interface RoadmapQuestion {
  id: string;
  tags: string[];
  introduction: string;
  question: string;
  options: string[];
  category: CategoryType;
  difficulty: 1 | 2 | 3 | 4 | 5;
  /** The version of this exact wording, with whatever review is recorded for
   * it. Absent means say nothing about review, not assume the best case. */
  review?: PublicItemReview;
}

// A playable lesson (a level) or exam (a checkpoint). `ref` is the level or
// checkpoint number; `passPct` is the score needed to clear it.
export interface RoadmapPlayable {
  kind: 'level' | 'checkpoint';
  topic: RoadmapTopic;
  ref: number;
  title: string;
  difficulty?: 1 | 2 | 3 | 4 | 5;
  passPct: number;
  sessionId: string;
  questions: RoadmapQuestion[];
  /** devShark code topics: the level's coding tasks, each with its own sealed session. */
  coding?: { task: PlayableCodingTask; session: string }[];
}

export interface RoadmapAnswerResult {
  selectedIndex: number;
  correctAnswer: number;
  isCorrect: boolean;
  explanation: string;
}

export interface RoadmapCompletionResult {
  correctAnswers: number;
  totalQuestions: number;
  percentage: number;
  passed: boolean;
  applied: boolean;
  /** Coding task ids the level still needs before it passes. */
  codingPending?: string[];
  /** The attempt was closed without a verdict because a question in it was
   * retired while it was open. Nothing was recorded; the level can be opened
   * again with its current questions. */
  invalidated?: { reason: 'content_retired'; questionIds: string[] };
  progress?: Partial<Record<RoadmapTopic, {
    levels: Record<string, { passed: boolean; bestPct: number }>;
    checkpoints: Record<string, { passed: boolean; bestPct: number }>;
  }>>;
}

File: mobile/lib/models/quiz.dart
class Question {
  final String id;
  final String question;
  final List<String> options;
  final String category;
  final int difficulty;
  final List<String> tags;
  final String? introduction;

  Question({
    required this.id,
    required this.question,
    required this.options,
    required this.category,
    required this.difficulty,
    required this.tags,
    this.introduction,
  });

  factory Question.fromJson(Map<String, dynamic> j) => Question(
        id: j['id'] as String,
        question: j['question'] as String,
        options: (j['options'] as List).cast<String>(),
        category: j['category'] as String,
        difficulty: (j['difficulty'] as num).toInt(),
        tags: ((j['tags'] as List?) ?? const <dynamic>[]).cast<String>(),
        introduction: j['introduction'] as String?,
      );
}

class QuizSession {
  final String sessionId;
  final List<Question> questions;
  QuizSession({required this.sessionId, required this.questions});

  factory QuizSession.fromJson(Map<String, dynamic> j) => QuizSession(
        sessionId: j['sessionId'] as String,
        questions: (j['questions'] as List)
            .map((q) => Question.fromJson(q as Map<String, dynamic>))
            .toList(),
      );
}

class QuestionResult {
  final String questionId;
  final int selectedIndex;
  final int correctAnswer;
  final bool isCorrect;
  final String explanation;

  QuestionResult({
    required this.questionId,
    required this.selectedIndex,
    required this.correctAnswer,
    required this.isCorrect,
    required this.explanation,
  });

  factory QuestionResult.fromJson(Map<String, dynamic> j) => QuestionResult(
        questionId: j['questionId'] as String,
        selectedIndex: (j['selectedIndex'] as num).toInt(),
        correctAnswer: (j['correctAnswer'] as num).toInt(),
        isCorrect: j['isCorrect'] as bool,
        explanation: (j['explanation'] as String?) ?? '',
      );
}

class QuizResult {
  final int totalQuestions;
  final int correctAnswers;
  final int percentage;
  final List<QuestionResult> results;

  QuizResult({
    required this.totalQuestions,
    required this.correctAnswers,
    required this.percentage,
    required this.results,
  });

  factory QuizResult.fromJson(Map<String, dynamic> j) => QuizResult(
        totalQuestions: (j['totalQuestions'] as num).toInt(),
        correctAnswers: (j['correctAnswers'] as num).toInt(),
        percentage: (j['percentage'] as num).toInt(),
        results: (j['results'] as List)
            .map((r) => QuestionResult.fromJson(r as Map<String, dynamic>))
            .toList(),
      );
}

const allCategories = <String>[
  'html',
  'css',
  'javascript',
  'typescript',
  'react',
  'nodejs',
  'git',
  'dev-world',
  'custom',
  'code-snippets',
];

const categoryLabels = <String, String>{
  'html': 'HTML',
  'css': 'CSS',
  'javascript': 'JavaScript',
  'typescript': 'TypeScript',
  'react': 'React',
  'nodejs': 'Node.js',
  'git': 'Git',
  'dev-world': 'Dev World',
  'custom': 'Custom',
  'code-snippets': 'Code Snippets',
};

const categoryColorsHex = <String, int>{
  'html': 0xFFE34C26,
  'css': 0xFF264DE4,
  'javascript': 0xFFF7DF1E,
  'typescript': 0xFF3178C6,
  'react': 0xFF61DAFB,
  'nodejs': 0xFF339933,
  'git': 0xFFF05032,
  'dev-world': 0xFF8B5CF6,
  'custom': 0xFF06B6D4,
  'code-snippets': 0xFFEC4899,
};

const difficultyModes = <Map<String, String>>[
  {'value': 'basics', 'label': 'Basics', 'tooltip': 'Definitions and basic terms.'},
  {'value': 'easy', 'label': 'Easy', 'tooltip': 'Difficulty 1-2.'},
  {'value': 'zero-to-hero', 'label': 'Zero to Hero', 'tooltip': 'Progressive 1 → 5.'},
  {'value': 'advanced', 'label': 'Advanced', 'tooltip': 'Difficulty 3-5.'},
  {'value': 'mixed', 'label': 'Mixed', 'tooltip': 'Random mix.'},
];
